cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -199,41 +211,84 @@ func (ks *kube2sky) addDNSUsingEndpoints(subdomain string, e *kapi.Endpoints) er
 	if err := ks.removeDNS(subdomain); err != nil {
 		return err
 	}
-	return ks.generateRecordsForHeadlessService(subdomain, e, svc)
+	return ks.generateRecordsForHeadlessService(subdomain, e, svc, isNewStyleFormat)
 }
 
 func (ks *kube2sky) handleEndpointAdd(obj interface{}) {
 	if e, ok := obj.(*kapi.Endpoints); ok {
 		name := buildDNSNameString(ks.domain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, false) })
 		name = buildDNSNameString(ks.domain, serviceSubdomain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, true) })
 	}
 }
 
-func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service) error {
+func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service, isNewStyleFormat bool) error {
+	etcd_key := subdomain
+	if isNewStyleFormat {
+		etcd_key = buildDNSNameString(subdomain, "a_rec")
+	}
 	for i := range service.Spec.Ports {
-		b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, service.Spec.Ports[i].Port))
+		port := service.Spec.Ports[i]
+		b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, port.Port))
 		if err != nil {
 			return err
 		}
-		glog.V(2).Infof("Setting DNS record: %v -> %q\n", subdomain, string(b))
-		if err := ks.writeSkyRecord(subdomain, string(b)); err != nil {
+		glog.V(2).Infof("Setting DNS record: %v -> %q, with etcd_key: %v\n", subdomain, string(b), etcd_key)
+		if err := ks.writeSkyRecord(etcd_key, string(b)); err != nil {
 			return err
 		}
+		if isNewStyleFormat {
+			portSegment := buildPortSegmentString(port.Name, port.Protocol)
+			if portSegment != "" {
+				err = ks.generateSRVRecord(subdomain, portSegment, "srv_rec", port.Port)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func buildPortSegmentString(portName string, portProtocol kapi.Protocol) string {
+	if portName == "" {
+		// we don't create a random name
+		return ""
+	}
+	portNameSegment := fmt.Sprintf("_%s", portName)
+
+	protocolSegment := string(portProtocol)
+	if protocolSegment == "" {
```

Comment:
you should never see this.  Protocol is defaulted in the API.  I would just log an error if you ever see this